Source/Core/VideoCommon/VertexLoaderARM64.cpp

```diff
@@ -0,0 +1,481 @@
+		LDR(INDEX_UNSIGNED, scratch2_reg, EncodeRegTo64(scratch2_reg), 0);
+		MUL(scratch1_reg, scratch1_reg, scratch2_reg);
+
+		MOVI2R(EncodeRegTo64(scratch2_reg), (u64)&cached_arraybases[array]);
+		LDR(INDEX_UNSIGNED, EncodeRegTo64(scratch2_reg), EncodeRegTo64(scratch2_reg), 0);
+		ADD(EncodeRegTo64(reg), EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch2_reg));
+	}
+}
+
+s32 VertexLoaderARM64::GetAddressImm(int array, u64 attribute, Arm64Gen::ARM64Reg reg, u32 align)
+{
+	if (attribute & MASK_INDEXED || (m_src_ofs & (align - 1)))
+		GetVertexAddr(array, attribute, reg);
+	else
+		return m_src_ofs;
+	return -1;
+}
+
+int VertexLoaderARM64::ReadVertex(u64 attribute, int format, int count_in, int count_out, bool dequantize, u8 scaling_exponent, AttributeFormat* native_format)
+{
+	static const float GC_ALIGNED16(scale_factors[]) =
+	{
+		1.0 / (1ULL <<  0), 1.0 / (1ULL <<  1), 1.0 / (1ULL <<  2), 1.0 / (1ULL <<  3),
+		1.0 / (1ULL <<  4), 1.0 / (1ULL <<  5), 1.0 / (1ULL <<  6), 1.0 / (1ULL <<  7),
+		1.0 / (1ULL <<  8), 1.0 / (1ULL <<  9), 1.0 / (1ULL << 10), 1.0 / (1ULL << 11),
+		1.0 / (1ULL << 12), 1.0 / (1ULL << 13), 1.0 / (1ULL << 14), 1.0 / (1ULL << 15),
+		1.0 / (1ULL << 16), 1.0 / (1ULL << 17), 1.0 / (1ULL << 18), 1.0 / (1ULL << 19),
+		1.0 / (1ULL << 20), 1.0 / (1ULL << 21), 1.0 / (1ULL << 22), 1.0 / (1ULL << 23),
+		1.0 / (1ULL << 24), 1.0 / (1ULL << 25), 1.0 / (1ULL << 26), 1.0 / (1ULL << 27),
+		1.0 / (1ULL << 28), 1.0 / (1ULL << 29), 1.0 / (1ULL << 30), 1.0 / (1ULL << 31),
+	};
+
+	ARM64Reg coords = count_in == 3 ? Q31 : D31;
+	ARM64Reg scale = count_in == 3 ? Q30 : D30;
+
+	int elem_size = 1 << (format / 2);
+	int load_bytes = elem_size * count_in;
+	elem_size <<= 3;
+
+	if (count_in == 1)
+		m_float_emit.LDR(elem_size, INDEX_UNSIGNED, coords, EncodeRegTo64(scratch1_reg), 0);
+	else
+		m_float_emit.LD1(elem_size, 1, coords, EncodeRegTo64(scratch1_reg));
+
+	if (format != FORMAT_FLOAT)
+	{
+		// Extend and convert to float
+		switch (format)
+		{
+		case FORMAT_UBYTE:
+			m_float_emit.UXTL(8, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+			m_float_emit.UXTL(16, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+		break;
+		case FORMAT_BYTE:
+			m_float_emit.SXTL(8, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+			m_float_emit.SXTL(16, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+		break;
+		case FORMAT_USHORT:
+			m_float_emit.REV16(8, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+			m_float_emit.UXTL(16, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+		break;
+		case FORMAT_SHORT:
+			m_float_emit.REV16(8, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+			m_float_emit.SXTL(16, EncodeRegToDouble(coords), EncodeRegToDouble(coords));
+		break;
+		}
+
+		m_float_emit.SCVTF(32, coords, coords);
+
+		if (dequantize && scaling_exponent)
+		{
+			MOVI2R(EncodeRegTo64(scratch2_reg), (u64)&scale_factors[scaling_exponent]);
```

Comment:
Probably right. We've got plenty of registers to spare, we could use it to point to the base of the scale_factors array.